### ISSUE TITLE
moment.ISO_8601 constant not accepted by format method

### ISSUE THREAD

**Issue by mj1856** [moment/moment#3019]
We have the `moment.ISO_8601` constant, but it only works when parsing.  It should also be accepted on formatting.

Currently, `moment().format(moment.ISO_8601)` gives `Uncaught TypeError: format.match is not a function`

**Comment by mj1856** [comment]
See also #3020.  Passing `moment.ISO_8601` should take this into account as well.  

Basically, `.format(moment.ISO_8601)` should be equivalent to `.format()`

**Comment by icambron** [comment]
See #2078 and #2375

**Comment by maggiepint** [comment]
Would it make sense to make the result of ``moment().format(moment.ISO_8601)`` not be equal to the result of ``moment().format()``, but instead the same as the result of ``moment().toISOString()``, except in local time?
I feel like this would help a lot of people who get stuck trying to use ``.toISOString()`` without realizing that it always converts to UTC. Probably what is going on is that they just want milliseconds.

So basically, implement as:
```js
moment().format(moment.ISO_8601) === moment().format('YYYY-MM-DDTHH:mm:ss.SSSZ')
```

**Comment by icambron** [comment]
Not a fan of that dichotomy -- seems weird that the format constant and the explicit method would do subtlety different things. I think that if we want to allow local time in `toISOString()` (and I do!) we should do that (see #1751).

Right now, that `ISO_8601` thing is an empty function, which is a little odd:

```js

    // constant that refers to the ISO standard
    utils_hooks__hooks.ISO_8601 = function () {};

    // date from string and format string
    function configFromStringAndFormat(config) {
        // TODO: Move this to another part of the creation flow to prevent circular deps
        if (config._f === utils_hooks__hooks.ISO_8601) {
            configFromISO(config);
            return;
        }
        //rest of configFromStringAndFormat
```

I guess it's just that on the parsing side, we have special-cased ISO logic, so you need some non-string flag to handle it. So to get the expected behavior in `format()`, you'd have to special case there too. I think it should just be:

```js
if (format == ISO_8601) {
  return this.toISOString();
}
```

**Comment by maggiepint** [comment]
I appreciate what you're saying about not having the format constant and the empty format do different things.

I'm not following you with your last code though.

```js
if (format == ISO_8601) {
  return this.toISOString();
}
```

That results in exactly what I was suggesting (if we fix toISOString), which you said isn't what you want.

Ultimately, I think what you wanted was to change the format function to read:

```js
export function format (inputString) {
    var output = formatMoment(this, (inputString === hooks.ISO_8601 || !inputString) ? hooks.defaultFormat : inputString);
    return this.localeData().postformat(output);
}
```

Right?

At this point, if that's what we want, this issue is done. I should write the unit tests and make the pull request :-)

**Comment by icambron** [comment]
All I was saying was for the `format == 'ISO_8601'` option to do the same thing as `toISOString()`, if it does anything at all.

**Comment by maggiepint** [comment]
Okay. I think I get it now.
You're saying that ``.format(moment.ISO_8601)`` should always delegate to ``.toISOString()``

That would make this change produce a date that is always UTC, unless we fixed toISOString.

It would be different than what Matt suggested in the original work item, and different than what I suggested as well.

**Comment by icambron** [comment]
Yup! Sorry for not being more clear.

**Comment by jough** [comment]
+1

**Comment by ilovett** [comment]
It's my understanding that `.format('YYYY-MM-DDTHH:mm:ss.SSSZ')` would produce a timezone relative output with milliseconds, EG: `2016-07-21T15:56:00.123-07:00`

Where `toISOString()` would produce UTC `2016-07-21T22:56:00.123Z`

**Comment by icambron** [comment]
@ilovett Correct

**Comment by mj1856** [comment]
Sorry I didn't see this sooner.

I strongly disagree with just directing it to `toISOString`.  It needs to honor the mode, offset, tz, etc. of the moment object.  I might do something like:

```javascript
moment.utc("2016-01-01T00:00:00", moment.ISO_8601, true).tz("Asia/Tokyo").format(moment.ISO_8601)

// Expected: ""2016-01-01T09:00:00+09:00""
// Not "2016-01-01T00:00:00Z"
```

**Comment by mj1856** [comment]
WRT `toISOString()` - we carry that over from the `Date` object's behavior.  We should perhaps just deprecate it?  It is weird to have two different methods for formatting output, one that honors mode and one that doesn't.

**Comment by icambron** [comment]
@mj1856 I think we actually agree. I'm just prioritizing the consistency higher: `toISOString()` and `format(ISO_8601)` should do the same thing; it's just *super* non-obvious that one would show the offset and the other not. If you want to change `toISOString()` to show the offset, I'm all for that! But the point remains that `ISO_8601` should delegate to `toISOString()`, regardless of what they do.

**Comment by ichernev** [comment]
We can release this in v3, then. Because its changing behavior for `toISOString`.

**Comment by marwahaha** [comment]
See https://momentjs.com/docs/#/-project-status/

I will try to avoid making any breaking changes to core Moment functionality.

If you need a constant, you can use one here: https://momentjs.com/docs/#/parsing/special-formats/

for example: `moment.HTML5_FMT.DATETIME_LOCAL_SECONDS`
